### ISSUE TITLE
GoogLeNet not training

### ISSUE THREAD

**Issue by korkinof** [soumith/imagenet-multiGPU.torch#2]
Hi guys,
I'm having some very weird issues. AlexNet is training fine, but when changing the model to GoogLeNet, the network just never seems to train.
I originally tried with a small dataset I already had and even after many epochs the classification result was still random.
Currently testing it with ImageNet and have the same problem. I am using batch size 60 and epoch size 10000. Of course I should probably wait for a few epochs, but after 1-2 epochs the Loss function error has not moved away from 6.9 and the accuracy is virtually 0.
I have also tried many different model implementations of GoogLeNet in case there is a problem with this particular one, but to no avail.
Any idea what might be the problem? Can it be that data augmentation is a prerequisite or something?

**Comment by soumith** [comment]
Add batchnorm after every convolution layer, your network will magically start training.

**Comment by soumith** [comment]
For an.example of adding batchnorm properly, use models/alexnetowtbn.lua as an example.

**Comment by korkinof** [comment]
Many thanks! I'll try this out!

**Comment by korkinof** [comment]
It worked like a charm!! Many many thanks!

**Comment by korkinof** [comment]
Hi guys,
SpatialBatchNormalisation indeed solved my issue and now I can get performance equivalent to the one I get on Caffe for my problem.
However, I cannot help it but wonder why was it so necessary! On Caffe the same network (without local response normalisation, or any normalisation for that matter) trains normally. Why can't Torch GoogLeNet do without BN and how was this network used before early 2015 when BN was proposed and implemented?

**Comment by soumith** [comment]
@korkinof BatchNorm is not necessary to get googlenet to train in torch. If you initialize the weights of each layer according to the paper, you will start learning in torch as well.

> how was this network used before early 2015 when BN was proposed and implemented?

By initializing weights properly.

Also, Caffe's googlenet implementation is kind of bad. It only goes upto 68% accuracy, vs the original googlenet paper goes to 72%.

**Comment by gcr** [comment]
@soumith I can't find details of GoogLeNet's weight initialization in the arXiv paper. Where does it talk about this? The only occurence of "initialization" is when they talk about how all of their networks in the ensemble were accidentally initialized to the same weights. Many thanks.

**Comment by korkinof** [comment]
Yes, I don't think initialisation is mentioned in the paper, but I think Caffe uses Xavier initialisation of the weights and I guess Torch doesn't (?).
You can take a look here: http://deepdish.io/2015/02/24/network-initialization/

**Comment by korkinof** [comment]
Hi guys,
I can verify that the source of this issue is indeed the initialisation. I tested with Xavier initialisation and without batch normalisation and the network trains, when previously it did not.
Initialising is very easy and you can see an example here:
https://github.com/e-lab/torch-toolbox/tree/master/Weight-init
Although the example needs some modification to work with inception layer.
In case you need my code, just ask!
Best wishes to all!

**Comment by gcr** [comment]
Wow, that is really helpful! Thanks, @korkinof .

Another huge important difference is that Caffe models from the model zoo expect the input to have range [0,255]. After mean subtraction, it should be roughly [-127, 127]. I'm using the `dp` library which scales images to have zero mean and unit variance, which causes my models not to converge. I'm not sure what the scripts in this repository do, but this difference is important because Caffe's style of input scaling effectively makes the gradients 255 times stronger than they otherwise would be. My Alexnet model didn't converge until I used the proper input scaling method. Just something else to look into.

**Comment by soumith** [comment]
the inputs being [-127, +127] is bad for many reasons, the simplest being that almost all the fancy weight initialization schemes are formulated with the assumptions that the inputs are [-1, 1] with zero mean. If you want stronger gradients, you can simply multiply learning rate.

**Comment by soumith** [comment]
and sorry for the late reply @gcr , the googlenet weight initialization , i got from the authors after emailing them. I remember some details in the original Googlenet paper, I have to open up the paper again and find it.

**Comment by gcr** [comment]
@soumith hm, interesting! with torch default weight initialization, my models don't converge with [-1, 1] input (ie. satisfying all the assumptions for picking default learning rates), but it does converge with [-255, 255] input scaling. I'll have to look at this more closely. Definitely a topic to study closer.

(full disclosure: i'm not using the code in this repository, but i'm not sure of any pertinent differences)

**Comment by gheinrich** [comment]
I am also having troubles getting GoogLeNet to converge. I am using a single GPU. I am using CIFAR10 dataset (upscaled to 3x224x224) and input data are normalized to approximately zero mean and unit variance. My batch size is 32. I have tried various learning rates between 0.1 and 0.001. I have tried Xavier initialization using @korkinof code (not sure if anything special needs doing for the inception layers). How many epochs do you typically need to wait before the network produces better results than random?  Thanks!

**Comment by korkinof** [comment]
You are using SpatialBatchNormalisation after the convolutional layers right?
According to my experience, with 0.01 learning rate and SpatialBatchNormalisation it starts training very quickly.
The Xavier initialisation made it possible to run Googlenet without BN, but it seemed to train slower. On the long run using Xavier init and BN did not make significant different for us.
Are you sure it's a good idea to upscale these tiny images so much? Why not try imagenet or a subset of it to see if your model is correct?

**Comment by gheinrich** [comment]
Thanks @korkinof. Actually no I didn't have batch normalisation layers as the other comments made it look like it was kind of hacky and I was under the impression that the preferred way was better initialization of weights. With SpatialBatchNormalisation layers (in the feature layers, I didn't add batch normalisation in the inception layers), the network is learning something pretty quickly. Thanks!

**Comment by arashno** [comment]
@korkinof 
I tried the Xavier weight initialization but still it does not converge.
Could you please give me your codes?
Thanks
